src/platforms/esp32/components/avm_builtins/i2c_resource.c

```diff
@@ -0,0 +1,743 @@
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    //
+    // ensure there is not a currently running transmission
+    //
+    if (UNLIKELY(!term_is_invalid_term(rsrc_obj->transmitting_pid))) {
+        ESP_LOGE(TAG, "nif_begin_transmission: Another process is in the process of transmitting.");
+
+        // {error, {einprogress, Pid :: pid()}}
+        if (UNLIKELY(memory_ensure_free(ctx, 2 * TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+            ESP_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+            return OUT_OF_MEMORY_ATOM;
+        }
+        term reason = create_pair(
+            ctx,
+            globalcontext_make_atom(global, EINPROGRESS_ATOMSTR),
+            rsrc_obj->transmitting_pid
+        );
+        return create_error_tuple(ctx, reason);
+    }
+
+    //
+    // extract the arguments
+    //
+
+    term address = argv[1];
+    uint8_t addr = term_to_int32(address);
+
+    //
+    // Initiate the I2C command
+    //
+
+    esp_err_t err;
+    i2c_cmd_handle_t cmd = i2c_cmd_link_create();
+    {
+        err = i2c_master_start(cmd);
+        CHECK_ERROR(ctx, err, "nif_write_bytes; enqueue start bit");
+
+        err = i2c_master_write_byte(cmd, (addr << 1) | I2C_MASTER_WRITE, 0x01);
+        CHECK_ERROR(ctx, err, "nif_write_bytes; enqueue I2C bus address");
+    }
+    rsrc_obj->transmitting_pid = term_from_local_process_id(ctx->process_id);
+    rsrc_obj->cmd = cmd;
+
+    return OK_ATOM;
+}
+
+//
+// i2c:enqueue_write_bytes_nif/1
+//
+
+static term nif_enqueue_write_bytes(Context *ctx, int argc, term argv[])
+{
+    TRACE("nif_enqueue_write_bytes\n");
+    UNUSED(argc);
+    GlobalContext *global = ctx->global;
+
+    //
+    // extract the resource
+    //
+    term i2c_resource = argv[0];
+    struct I2CResource *rsrc_obj;
+    if (UNLIKELY(!to_i2c_resource(i2c_resource, &rsrc_obj, ctx))) {
+        ESP_LOGE(TAG, "Failed to convert i2c_resource");
+        RAISE_ERROR(BADARG_ATOM);
+    }
+
+    //
+    // ensure there the currently running transmission is the current process
+    //
+    term pid = term_from_local_process_id(ctx->process_id);
+    if (UNLIKELY(rsrc_obj->transmitting_pid != pid)) {
+        ESP_LOGE(TAG, "nif_end_transmission: Another process is in the process of transmitting.");
+
+        // {error, {einprogress, Pid :: pid()}}
+        if (UNLIKELY(memory_ensure_free(ctx, 2 * TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+            ESP_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+            return OUT_OF_MEMORY_ATOM;
+        }
+        term reason = create_pair(
+            ctx,
+            globalcontext_make_atom(global, EINPROGRESS_ATOMSTR),
+            rsrc_obj->transmitting_pid
+        );
+        return create_error_tuple(ctx, reason);
+    }
+
+    term data = argv[1];
+
+    esp_err_t err;
+    err = i2c_master_write(rsrc_obj->cmd, (unsigned char *) term_binary_data(data), term_binary_size(data), true);
```

Comment:
I'm worried that if GC runs after this function, we might end sending garbage, since the area pointed by `term_binary_data(data)` might be freed and reused for other purposes.

esp-idf doc says: `This buffer shall remain valid until the transaction is finished.`

Reply:
Yet another reason not to support the begin_transmission/end_tranmission API, which is an API in search of a use-case.  As you can see, this is also an issue for the port-based implementation.  So I would suggest dropping that particular API (or force users to only write a byte at a time)

Reply:
Good catch.  Changed to use `i2c_master_write_byte` and write one byte at a time.